signac/core/validators.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) 2017 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Validators for SyncedCollection API."""
+import warnings
+from collections.abc import Mapping
+from collections.abc import Sequence
+
+from ..errors import InvalidKeyError
+from ..errors import KeyTypeError
+from .synced_collection import NUMPY
+
+if NUMPY:
+    import numpy
+
+
+def NoDotInKey(data):
+    """Raise an exception if there is dot in key.
+
+    Parameters
+    ----------
+    data:
+        Data to validate.
+
+    Returns
+    -------
+    data
+
+    Raises
+    ------
+    KeyTypeError:
+        If keys have unsupported data type.
+    InvalidKeyError
+        If key is not supported.
+    """
+    VALID_KEY_TYPES = (str, int, bool, type(None))
+
+    if isinstance(data, Mapping):
+        for key in data.keys():
+            if isinstance(key, str):
+                if '.' in key:
+                    raise InvalidKeyError(
+                        "Mapping keys may not contain dots ('.'): {}".format(key))
+            elif not isinstance(key, VALID_KEY_TYPES):
+                raise KeyTypeError(
+                    "Mapping keys must be str, int, bool or None, not {}"
+                    .format(type(key).__name__))
+    return data
+
+
+def JSONFormatValidator(data):
+    """Implement the validation for JSON serializable data
+
+    Parameters
+    ----------
+    data:
+        Data to validate.
+
+    Returns
+    -------
+    data
+
+    Raises
+    ------
+    KeyTypeError:
+        If keys of mapping have unsupported data type.
+    TypeError
+        If data type is not supported.
+    """
+
+    if isinstance(data, (str, int, float, bool, bytes, type(None))):
```

Comment:
I don't think `bytes` is JSON-serializable -- this could be a bug. I wonder if this is a holdover from Python 2, when `str` and `bytes` were handled differently?

```pytb
>>> some_bytes = b'123'
>>> import json
>>> json.dumps(some_bytes)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/bdice/miniconda3/envs/dice/lib/python3.8/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/home/bdice/miniconda3/envs/dice/lib/python3.8/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/home/bdice/miniconda3/envs/dice/lib/python3.8/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/home/bdice/miniconda3/envs/dice/lib/python3.8/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type bytes is not JSON serializable
>
```

Reply:
Good catch. That certainly seems likely, and I'm fine just disallowing bytes.